File: integrations/hermes/websidian/sites.py
"""Vault -> Websidian site mapping shared by the agent plugin and the dashboard extension. Standard library only.

Both processes (the Hermes agent/gateway loading ``__init__.py`` and the dashboard importing
``dashboard/plugin_api.py``) derive site slugs from the same ``vaults`` setting with :func:`normalize_vaults`,
so the links the agent shares point at the sites the dashboard serves.
"""

from __future__ import annotations

import base64
import os
import re
from typing import Any, Dict, Iterable, List, Mapping, Optional
from urllib.parse import parse_qs, quote, urlsplit

# Websidian is mounted at this basePath (after any dashboard URL prefix) inside the dashboard.
DASHBOARD_TAB_PATH = "/websidian"
WEBSIDIAN_MOUNT = "/api/plugins/websidian/w"

DEFAULT_PORT = 8095
DEFAULT_PUBLIC_BASE = "http://localhost:9119"
LINK_STYLES = ("dashboard", "direct")

# Deep-link query parameters of the dashboard tab. The ``64`` ones carry the same value as base64url;
# see :func:`note_query`. ``dashboard/dist/index.js`` reads and writes both.
NOTE_PARAM, NOTE_B64_PARAM = "note", "note64"
QUERY_PARAM, QUERY_B64_PARAM = "q", "q64"

_URI_COMPONENT_SAFE = "!*'()"
# Characters a percent-decode never changes (unreserved plus the query punctuation an iframe query uses).
# "%", "&", "+", "#" and anything needing escaping are deliberately absent.
_DECODE_STABLE = re.compile(r"[A-Za-z0-9\-._~!*'()=,:/]*\Z")


def _enc(value: str) -> str:
    return quote(str(value), safe=_URI_COMPONENT_SAFE, encoding="utf-8", errors="surrogatepass")


def as_bool(value: Any, default: bool) -> bool:
    if isinstance(value, bool):
        return value
    if value is None:
        return default
    s = str(value).strip().lower()
    if s in ("1", "true", "yes", "on"):
        return True
    if s in ("0", "false", "no", "off"):
        return False
    return default


def slugify(value: str) -> str:
    """Lowercase ``[a-z0-9_-]``; never starts with ``_`` (Websidian's own routes: ``/_health``, ``/_static``)."""
    s = re.sub(r"[^a-z0-9_-]+", "-", str(value or "").strip().lower())
    s = re.sub(r"-{2,}", "-", s).strip("-_")
    return s[:64].strip("-_")


def _url_slug(url: str) -> str:
    try:
        path = urlsplit(str(url)).path
    except ValueError:
        return ""
    segs = [s for s in path.split("/") if s]
    return slugify(segs[-1]) if segs else ""


def _folder_slug(path: str) -> str:
    p = str(path or "").replace("\\", "/").rstrip("/")
    return slugify(p.rsplit("/", 1)[-1]) if p else ""


def normalize_vaults(vaults: Optional[Iterable[Any]]) -> List[Dict[str, Any]]:
    """``[{path, url, slug, title, edit, untrusted}]`` with unique slugs, in the configured order.

    slug: explicit ``slug`` -> last path segment of ``url`` -> folder name -> ``vault``; duplicates get ``-2``, ``-3``...
    ``untrusted`` defaults to True and ``edit`` to False: a vault an agent writes to is also a vault every
    dashboard user could rewrite, so browser editing is opt-in per vault.
    """
    out: List[Dict[str, Any]] = []
    used: set = set()
    for v in vaults or ():
        if not isinstance(v, Mapping):
            continue
        path = str(v.get("path") or "").strip()
        if not path:
            continue
        url = str(v.get("url") or "").strip()
        if url and not url.endswith("/"):
            url += "/"
        base = slugify(str(v.get("slug") or "")) or _url_slug(url) or _folder_slug(path) or "vault"
        slug, n = base, 2
        while slug in used:
            slug = f"{base}-{n}"
            n += 1
        used.add(slug)
        title = str(v.get("title") or "").strip() or os.path.basename(path.replace("\\", "/").rstrip("/")) or slug
        out.append({
            "path": path, "url": url, "slug": slug, "title": title,
            # Off unless the operator chose this vault deliberately: every dashboard user shares the editor.
            "edit": as_bool(v.get("edit"), False),
            # Never let the default be False: vault text may come from the agent, web pages, tool output.
            "untrusted": as_bool(v.get("untrusted"), True),
        })
    return out


def public_prefix(public_base: str) -> str:
    """URL path prefix of the dashboard (``https://x.example/hermes`` -> ``/hermes``), no trailing slash."""
    try:
        path = urlsplit(str(public_base or "")).path
    except ValueError:
        return ""
    path = "/" + "/".join(s for s in path.split("/") if s)
    return "" if path == "/" else path


def dashboard_settings(raw: Any) -> Dict[str, Any]:
    """Normalized ``dashboard`` setting: ``{configured, port, app_dir, node, public_base}`` (``app_dir`` may be '')."""
    d = raw if isinstance(raw, Mapping) else {}
    try:
        port = int(d.get("port") or DEFAULT_PORT)
    except (TypeError, ValueError):
        port = DEFAULT_PORT
    if not 1 <= port <= 65535:
        port = DEFAULT_PORT
    public_base = str(d.get("public_base") or "").strip().rstrip("/") or DEFAULT_PUBLIC_BASE
    return {
        "configured": isinstance(raw, Mapping),
        "port": port,
        "app_dir": str(d.get("app_dir") or "").strip(),
        "node": str(d.get("node") or "node").strip() or "node",
        "public_base": public_base,
    }


def websidian_base_path(public_base: str) -> str:
    return public_prefix(public_base) + WEBSIDIAN_MOUNT


def direct_site_url(public_base: str, slug: str) -> str:
    """Websidian site base behind the dashboard: ``<public_base>/api/plugins/websidian/w/<slug>/``."""
    return str(public_base).rstrip("/") + WEBSIDIAN_MOUNT + "/" + slug + "/"


def note_rel(rel: str) -> str:
    """``Folder\\My Note.md`` -> ``Folder/My Note`` (forward slashes, no surrounding slashes, ``.md`` dropped)."""
    rel = str(rel).replace("\\", "/").strip("/")
    return rel[:-3] if rel.lower().endswith(".md") else rel


def note_param(rel: str) -> str:
    """``Folder/My Note.md`` -> ``Folder/My%20Note`` (each segment encodeURIComponent-ed, ``.md`` dropped)."""
    return "/".join(_enc(seg) for seg in note_rel(rel).split("/"))


def b64_param(value: str) -> str:
    """``value`` as base64url without padding: only ``[A-Za-z0-9_-]``, which percent-decoding never changes."""
    return base64.urlsafe_b64encode(str(value).encode("utf-8", "surrogatepass")).decode("ascii").rstrip("=")


def decode_b64_param(value: str) -> str:
    """Inverse of :func:`b64_param` (``""`` when ``value`` is not base64url)."""
    s = str(value or "")
    try:
        return base64.urlsafe_b64decode((s + "=" * (-len(s) % 4)).encode("ascii")).decode("utf-8", "surrogatepass")
    except ValueError:  # binascii.Error and UnicodeError are both ValueError
        return ""


def note_query(rel: str) -> str:
    """The ``note=``/``note64=`` part of a deep link for ``rel``.

    A deep link opened without a session goes through the login page, and Hermes decodes the target one
    time more than it encoded it (the gate percent-encodes ``path?query`` into ``/login?next=``, the HTTP
    layer decodes that query value, and ``_validate_post_login_target`` unquotes it again). So a value is
    only safe there when no percent-escape has to survive: names that need none keep the readable plain
    form, the rest travel as base64url in ``note64``. Both forms work when already signed in.
    """
    norm = note_rel(rel)
    plain = note_param(norm)
    return f"{NOTE_PARAM}={plain}" if plain == norm else f"{NOTE_B64_PARAM}={b64_param(norm)}"


def search_query(q: str) -> str:
    """The ``q=``/``q64=`` part of a deep link (the iframe's own query string); see :func:`note_query`."""
    q = str(q or "").lstrip("?")
    return f"{QUERY_PARAM}={q}" if _DECODE_STABLE.match(q) else f"{QUERY_B64_PARAM}={b64_param(q)}"


def _from_query(query: str, param: str, b64_param_name: str) -> str:
    """Value of ``param`` in a deep link's query string, preferring the base64url form. Reference
    implementation of what ``dashboard/dist/index.js`` (``readQuery``) does in the browser."""
    params = parse_qs(str(query or "").lstrip("?"), keep_blank_values=True)
    if params.get(b64_param_name):
        return decode_b64_param(params[b64_param_name][0])
    return params.get(param, [""])[0]


def note_from_query(query: str) -> str:
    """The note a deep link points at (``note64`` wins over the older plain ``note``)."""
    return _from_query(query, NOTE_PARAM, NOTE_B64_PARAM)


def search_from_query(query: str) -> str:
    """The iframe query string a deep link carries (``q64`` wins over the older plain ``q``)."""
    return _from_query(query, QUERY_PARAM, QUERY_B64_PARAM)


def dashboard_note_url(public_base: str, slug: str, rel: str, edit: bool = False) -> str:
    """``<public_base>/websidian?site=<slug>&note=<rel without .md>[&edit=1]`` (the dashboard tab deep link);
    the note travels as ``note64=<base64url>`` when its name would not survive the login redirect."""
    url = f"{str(public_base).rstrip('/')}{DASHBOARD_TAB_PATH}?site={_enc(slug)}&{note_query(rel)}"
    return url + "&edit=1" if edit else url


def resolve_links(vaults: List[Dict[str, Any]], link_style: Any, dashboard: Mapping[str, Any]) -> List[Dict[str, Any]]:
    """Add ``style`` (``dashboard`` | ``direct`` | ``""``), ``public_base`` and the direct ``url`` to each vault.

    - ``link_style: dashboard`` -> dashboard tab links on ``dashboard.public_base``.
    - ``link_style: direct`` -> the vault's ``url``, else ``<public_base>/api/plugins/websidian/w/<slug>/``.
    - unset: a vault with an explicit ``url`` keeps direct links to it (the original behaviour); without ``url``,
      dashboard links when the ``dashboard`` setting exists, else no links.
    """
    style_setting = str(link_style or "").strip().lower()
    style_setting = style_setting if style_setting in LINK_STYLES else ""
    public_base = dashboard.get("public_base") or DEFAULT_PUBLIC_BASE
    out = []
    for v in vaults:
        v = dict(v)
        if style_setting == "dashboard":
            v["style"] = "dashboard"
        elif style_setting == "direct":
            v["style"] = "direct"
            v["url"] = v.get("url") or direct_site_url(public_base, v["slug"])
        elif v.get("url"):
            v["style"] = "direct"
        elif dashboard.get("configured"):
            v["style"] = "dashboard"
        else:
            v["style"] = ""
        v["public_base"] = public_base
        out.append(v)
    return out

File: integrations/openclaw/websidian/lib/sites.js
// Vault -> Websidian site mapping and plugin settings. Pure, standard library only.
//
// The agent hooks, the /brain command, the websidian_links tool and the pages served behind the Gateway
// all derive site slugs from the same `vaults` setting through normalizeVaults(), so the links the agent
// shares point at the sites the plugin serves.
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

// Websidian is mounted at this path on the Gateway: /plugins/websidian/w/<slug>/...
export const ROUTE_PREFIX = '/plugins/websidian';
export const WEBSIDIAN_MOUNT = ROUTE_PREFIX + '/w';
// The native Memory surface in OpenClaw's Control UI: a second HTTP route, authenticated by the Gateway
// itself (auth: "gateway"), so the operator signs in once — to OpenClaw — and never again here.
//
// A sibling of the stand-alone prefix, not a child of it: OpenClaw refuses two plugin routes whose
// prefixes overlap ("http route overlap rejected", seen on 2026.9.5), and the stand-alone pages must keep
// their own path and their own sign-in. The session cookie the Memory route mints is still scoped to
// ROUTE_PREFIX, so it reaches the proxy and never travels to this route.
export const MEMORY_PREFIX = ROUTE_PREFIX + '-memory';
export const MEMORY_PAGE_ID = 'memory';
export const DEFAULT_PORT = 8095;
export const DEFAULT_GATEWAY_PORT = 18789;
export const DEFAULT_SESSION_HOURS = 12;

export const DEFAULT_PROTECT = Object.freeze([
  'SKILL.md', 'SOUL.md', 'AGENTS.md', 'MEMORY.md', 'USER.md', 'TOOLS.md', 'IDENTITY.md', 'HEARTBEAT.md', 'BOOTSTRAP.md',
]);

export function asBool(value, fallback) {
  if (typeof value === 'boolean') return value;
  if (value === undefined || value === null) return fallback;
  const s = String(value).trim().toLowerCase();
  if (['1', 'true', 'yes', 'on'].includes(s)) return true;
  if (['0', 'false', 'no', 'off'].includes(s)) return false;
  return fallback;
}

// Lowercase [a-z0-9_-]; never starts with "_" (Websidian's own routes: /_health, /_static).
export function slugify(value) {
  let s = String(value || '').trim().toLowerCase().replace(/[^a-z0-9_-]+/g, '-');
  s = s.replace(/-{2,}/g, '-').replace(/^[-_]+|[-_]+$/g, '');
  return s.slice(0, 64).replace(/^[-_]+|[-_]+$/g, '');
}

function urlSlug(url) {
  let pathname;
  try { pathname = new URL(String(url)).pathname; } catch { return ''; }
  const segs = pathname.split('/').filter(Boolean);
  return segs.length ? slugify(segs[segs.length - 1]) : '';
}

function folderSlug(p) {
  const s = String(p || '').replace(/\\/g, '/').replace(/\/+$/, '');
  return s ? slugify(s.split('/').pop()) : '';
}

// [{path, url, slug, title, edit, untrusted}] with unique slugs, in the configured order.
// slug: explicit -> last path segment of url -> folder name -> "vault"; duplicates get -2, -3...
// untrusted defaults to true and edit to false: a vault an agent writes to is also a vault every browser
// user could rewrite, so browser editing is opt-in per vault.
export function normalizeVaults(vaults) {
  const out = [];
  const used = new Set();
  for (const v of Array.isArray(vaults) ? vaults : []) {
    if (!v || typeof v !== 'object') continue;
    const p = String(v.path || '').trim();
    if (!p) continue;
    let url = String(v.url || '').trim();
    if (url && !url.endsWith('/')) url += '/';
    const base = slugify(v.slug || '') || urlSlug(url) || folderSlug(p) || 'vault';
    let slug = base;
    for (let n = 2; used.has(slug); n++) slug = `${base}-${n}`;
    used.add(slug);
    const title = String(v.title || '').trim() || path.basename(p.replace(/\\/g, '/').replace(/\/+$/, '')) || slug;
    out.push({ path: p, url, slug, title, edit: asBool(v.edit, false), untrusted: asBool(v.untrusted, true) });
  }
  return out;
}

// URL path prefix of the Gateway as browsers see it ("https://x.example/openclaw" -> "/openclaw"), no trailing slash.
export function publicPrefix(publicBase) {
  let pathname;
  try { pathname = new URL(String(publicBase || '')).pathname; } catch { return ''; }
  const p = '/' + pathname.split('/').filter(Boolean).join('/');
  return p === '/' ? '' : p;
}

export function websidianBasePath(publicBase) {
  return publicPrefix(publicBase) + WEBSIDIAN_MOUNT;
}

// Site base behind the Gateway: <publicBase>/plugins/websidian/w/<slug>/
export function siteUrl(publicBase, slug) {
  return String(publicBase).replace(/\/+$/, '') + WEBSIDIAN_MOUNT + '/' + slug + '/';
}

export function expandHome(p, env = process.env) {
  const s = String(p || '');
  if (s === '~' || s.startsWith('~/') || s.startsWith('~\\')) {
    const home = env.HOME || env.USERPROFILE || os.homedir();
    return path.join(home, s.slice(1));
  }
  return s;
}

// OpenClaw's state directory: $OPENCLAW_STATE_DIR, else ~/.openclaw (mirrors OpenClaw's own default).
export function resolveStateDir(env = process.env) {
  const configured = String(env.OPENCLAW_STATE_DIR || '').trim();
  if (configured) return path.resolve(expandHome(configured, env));
  const home = env.HOME || env.USERPROFILE || os.homedir();
  return path.join(home, '.openclaw');
}

// The agent workspaces the Gateway config names: agents.defaults.workspace and every agents.list[].workspace,
// plus OpenClaw's default (<state dir>/workspace, or workspace-<profile>) when nothing is configured.
export function workspaceDirs(cfg, env = process.env) {
  const out = new Set();
  const agents = (cfg && cfg.agents) || {};
  const add = (p) => { if (typeof p === 'string' && p.trim()) out.add(path.resolve(expandHome(p.trim(), env))); };
  add(agents.defaults && agents.defaults.workspace);
  for (const a of Array.isArray(agents.list) ? agents.list : []) add(a && a.workspace);
  if (!(agents.defaults && agents.defaults.workspace)) {
    const profile = String(env.OPENCLAW_PROFILE || '').trim();
    out.add(path.join(resolveStateDir(env), profile && profile !== 'default' ? `workspace-${profile}` : 'workspace'));
  }
  return [...out];
}

// The workspace file tools resolve relative paths against for `agentId` (mirrors resolveAgentWorkspaceDir).
export function workspaceFor(cfg, agentId, env = process.env) {
  const agents = (cfg && cfg.agents) || {};
  const list = Array.isArray(agents.list) ? agents.list : [];
  const id = String(agentId || '').trim().toLowerCase();
  const own = list.find(a => a && String(a.id || '').toLowerCase() === id && typeof a.workspace === 'string' && a.workspace.trim());
  if (own) return path.resolve(expandHome(own.workspace.trim(), env));
  if (agents.defaults && typeof agents.defaults.workspace === 'string' && agents.defaults.workspace.trim()) {
    return path.resolve(expandHome(agents.defaults.workspace.trim(), env));
  }
  return workspaceDirs(cfg, env)[0];
}

// The Websidian runtime a packed plugin carries inside itself (deploy/pack.mjs), so `openclaw plugins install`
// alone is a complete install: on first start the supervisor copies it to <dataDir>/app and installs its locked
// dependencies there (supervisor.provision). It never runs in place: OpenClaw overrides some dependency
// versions for every plugin (path-to-regexp 8, which Express 4 cannot use) and loads plugins from a rebuilt
// copy. A plugin copied from a checkout by the installers has no runtime/; the installers fill <dataDir>/app.
export const BUNDLED_RUNTIME = path.resolve(fileURLToPath(new URL('../runtime/', import.meta.url)));
export function bundledRuntime(dir = BUNDLED_RUNTIME) {
  return fs.existsSync(path.join(dir, 'src', 'server.js')) ? dir : '';
}

// The vaults used when `vaults` is not set at all: the default agent's workspace, once it exists, so a fresh
// install shows the agent's memory without any config. An explicit `vaults: []` means none.
export function defaultVaults(cfg = {}, env = process.env) {
  const ws = workspaceFor(cfg, '', env);
  return ws && fs.existsSync(ws) ? [{ path: ws, slug: 'workspace', title: 'Agent workspace' }] : [];
}

// Normalized `ui` setting: {enabled, port, appDir, bundleDir, dataDir, node, publicBase, auth, password, sessionHours}.
export function uiSettings(raw, cfg = {}, env = process.env) {
  const d = raw && typeof raw === 'object' ? raw : {};
  let port = parseInt(d.port, 10);
  if (!Number.isInteger(port) || port < 1 || port > 65535) port = DEFAULT_PORT;
  const gatewayPort = parseInt(cfg && cfg.gateway && cfg.gateway.port, 10);
  const publicBase = String(d.publicBase || '').trim().replace(/\/+$/, '')
    || `http://127.0.0.1:${Number.isInteger(gatewayPort) && gatewayPort > 0 ? gatewayPort : DEFAULT_GATEWAY_PORT}`;
  const stateDir = resolveStateDir(env);
  const dataDir = String(d.dataDir || '').trim() ? path.resolve(expandHome(String(d.dataDir).trim(), env)) : path.join(stateDir, 'plugin-data', 'websidian');
  const appDirSet = !!String(d.appDir || '').trim();
  const appDir = appDirSet ? path.resolve(expandHome(String(d.appDir).trim(), env)) : path.join(dataDir, 'app');
  const auth = String(d.auth || '').trim().toLowerCase() === 'password' ? 'password' : 'gateway';
  const mem = d.memory && typeof d.memory === 'object' ? d.memory : {};
  let sessionHours = Number(d.sessionHours);
  if (!Number.isFinite(sessionHours) || sessionHours <= 0) sessionHours = DEFAULT_SESSION_HOURS;
  return {
    enabled: asBool(d.enabled, true),
    port,
    appDir,
    // Where provision() installs the runtime from: the bundled copy, unless ui.appDir names a runtime of its own.
    bundleDir: appDirSet ? '' : bundledRuntime(),
    dataDir,
    node: String(d.node || '').trim() || process.execPath,
    publicBase,
    auth,
    password: typeof d.password === 'string' ? d.password : '',
    sessionHours,
    // The Memory page: which vault it reads, and how it is labelled in the Control UI sidebar.
    memory: {
      enabled: asBool(mem.enabled, true),
      vault: String(mem.vault || '').trim(),
      label: String(mem.label || '').trim() || 'Memory',
      icon: String(mem.icon || '').trim() || 'brain',
      order: Number.isFinite(Number(mem.order)) ? Number(mem.order) : 20,
    },
  };
}

// Add `url` (the site base links point at) to each vault: an explicit external url wins; otherwise the site
// the plugin serves itself (when the pages are enabled); otherwise "" (no links).
// A plugin-served site with `edit: false` gets no edit link (Websidian would refuse the editor); an external
// url is somebody else's Websidian, whose config decides, so it keeps both.
export function resolveLinks(vaults, ui) {
  return vaults.map(v => {
    const out = { ...v };
    if (v.url) {
      out.external = true;
    } else if (ui && ui.enabled) {
      out.url = siteUrl(ui.publicBase, v.slug);
      out.external = false;
    } else {
      out.url = '';
      out.external = false;
    }
    return out;
  });
}
